### DISCUSSION TITLE
Wrapping code that depends on __cplusplus

### DISCUSSION THREAD

**Discussion by maleadt** [JuliaInterop/Clang.jl#391] [Q&A] [answered]
Another question relating to nasty CUDA headers 😅 

I'm trying to wrap headers that make use of the `half` datatype, which is only really defined in C++. Consequently, these functions are guarded by an `ifdef __cplusplus` check:

```
#if defined(__cplusplus)
CUBLASAPI cublasStatus_t CUBLASWINAPI cublasHSHgemvBatched(cublasHandle_t handle,
                                                           cublasOperation_t trans,
                                                           int m,
                                                           int n,
                                                           const float* alpha, /* host or device pointer */
                                                           const __half* const Aarray[],
                                                           int lda,
                                                           const __half* const xarray[],
                                                           int incx,
                                                           const float* beta, /* host or device pointer */
                                                           __half* const yarray[],
                                                           int incy,
                                                           int batchCount);
#endif
```

Now, given that these are _pointers_ to half, we can use these APIs from Julia and pass a `Ptr{Float16}`. Is there a way that I can have Clang.jl wrap these? Adding `__cplusplus` to the `defines` breaks all kinds of things, because in other headers that then introduces actual C++ code which we can't handle.

I know this is a bit of a long shot, but I'd like to avoid having to pre-patch the headers before processing them with Clang.jl (either manually, or with some `patch` invocations), so I'm curious about possible workarounds. Maybe it's possible to have a `define` for a single header file? or have a pre-processing step where I can have Julia code filter out these specific macros (if I detect that there's headers inside, and not other kinds of code)?

**Comment by Gnimuc** [comment, marked as answer]
That code snippet is dropped by C's preprocessor, so Clang's ASTUnit can't see it. As a result, we have no choice but to add C++ support in `Clang.Generators`, which needs a lot of refactoring work.

> Maybe it's possible to have a `define` for a single header file? 

As long as the header is a top-level independent header and can be parsed successfully(either being parsed standalone or using `-isystem` tricks), you can create a separate context with dedicated compilation flags.

**Reply by maleadt** [reply]
Sadly, it includes other headers, some of which become problematic in C++ mode.

**Reply by maleadt** [reply]
I'll do some preprocessing in Julia then; I could rewrite the headers to a temp file and parse that with Clang.jl instead.

**Comment by Gnimuc** [comment]
Wait. Is `__half` really a C++ thing?

https://stackoverflow.com/questions/73873870/is-the-half-a-native-c-type
